framework/src/bundle/BundleRegistry.cpp

```diff
@@ -62,36 +62,170 @@ void BundleRegistry::Clear()
     std::vector<std::shared_ptr<BundlePrivate>> alreadyInstalled;
-    while (range.first != range.second) {
-      auto b = range.first->second;
-      alreadyInstalled.push_back(b);
-      auto bu = coreCtx->bundleHooks.FilterBundle(
-        MakeBundleContext(b->bundleContext.Load()), MakeBundle(b));
-      if (bu)
-        res.push_back(bu);
-      ++range.first;
-    }
+    // Populate the res and alreadyInstalled vectors with the appropriate data
+    // based on what bundles are already installed
+    GetAlreadyInstalledBundlesAtLocation(range, res, alreadyInstalled);
 
+    // Perform the install
     auto newBundles = Install0(location, alreadyInstalled, caller);
     res.insert(res.end(), newBundles.begin(), newBundles.end());
     if (res.empty()) {
       throw std::runtime_error("All bundles rejected by a bundle hook");
     } else {
       return res;
     }
+  } else {
+    // Check to see if another thread is installing the currently uninstalled bundle
+    auto p = initialBundleInstallMap.find(location);
+
+    /*
+      If no other thread is installing the desired bundle, create a map entry,
+      install the bundle, and notify all other threads wanting to install this bundle
+      that it is safe to do so. If the current thread is not the installing thread, then
+      it will increment the reference count in the initialBundleInstallMap so that the map
+      entry isn't prematurely deleted, wait to be notified that the install thread is done,
+      and then perform the regular install procedure.
+    */
+    if (p == initialBundleInstallMap.end()) {
+      // Insert entry into the initialBundleInstallMap to prevent other threads from
+      // trying to install this uninstalled bundle at the same time
+      auto pairToInsert = std::make_pair(uint32_t(1), WaitCondition{});
+      initialBundleInstallMap.insert(
+        std::make_pair(location, std::move(pairToInsert)));
+
+      l.UnLock();
+
+      // Perform the install
+      auto installedBundles = Install0(location, {}, caller);
+      {
+        // Notify all waiting threads that it is safe to install the bundle
+        std::lock_guard<std::mutex> lock(
+          *(initialBundleInstallMap[location].second.m));
+        initialBundleInstallMap[location].second.waitFlag = false;
+        initialBundleInstallMap[location].second.cv->notify_all();
+      }
+
+      DecrementInitialBundleMapRef(l, location);
+
+      return installedBundles;
+    } else {
+      initialBundleInstallMap[location].first++;
+      l.UnLock();
+
+      {
+        // Wait for the install thread to notify this thread that it is safe
+        // to install the current bundle
+        std::unique_lock<std::mutex> lock(
+          *(initialBundleInstallMap[location].second.m));
+        initialBundleInstallMap[location].second.cv->wait(
+          lock, [&location, this] {
+            return !initialBundleInstallMap[location].second.waitFlag;
+          });
+
+        lock.unlock();
```

Comment:
A std::unique_lock is required. https://en.cppreference.com/w/cpp/thread/condition_variable/wait

Reply:
If I'm reading https://en.cppreference.com/w/cpp/thread/unique_lock/~unique_lock correctly, ```lock.unlock()``` is redundant since destruction of the ```std::unqiue_lock``` object will unlock the mutex.